File: Android App - Shredsheets 2/app/src/main/java/revert/factories/Factory.java
package revert.factories;

import java.util.Stack;

public abstract class Factory<T> {

    protected Stack<T> items;

    protected Factory() {
        items = new Stack<>();
        for (int i = 0; i < getEstimatedRequired(); i++)
            items.push(createItem());
    }


    public T get()
    {
        if (items.empty())
            for (int i = 0; i < getEstimatedRequired(); i++)
            {
                items.push(createItem());
            }

        return items.pop();
    }

    public void put(T item)
    {
        items.push(cleanItem(item));
    }

    protected abstract int getEstimatedRequired();

    protected abstract T createItem();

    protected abstract T cleanItem(T item);
}

File: Shredsheets 3/Kommon/src/main/java/com/bm/factories/Factory.kt
package com.bm.factories

import java.util.LinkedList
import java.util.Queue

abstract class Factory<T> protected constructor() {
    private var stack: Queue<T> = LinkedList()
    protected open val initialCapacity: Int = 5000

    init {
        for (i in 0 until initialCapacity) {
            val newItem = getNewItem()
            stack.add(newItem)
        }
    }

    protected abstract fun getNewItem(): T

    fun get(): T {
        return if (stack.isEmpty()) {
            getNewItem()
        } else stack.remove()
    }

    fun peek(): T {
        if (stack.isEmpty()) {
            stack.add(getNewItem())
        }
        return stack.peek()
    }

    open fun dispose(item: T) {
        stack.add(item)
    }

    open fun dispose(items: Collection<T>) {
        for (item in items) stack.add(item)
    }
}
